### DISCUSSION TITLE
Rancher Desktop

### DISCUSSION THREAD

**Discussion by enrico23** [rancher-sandbox/rancher-desktop#5437] [Q&A]
I started using Rancher Desktop as a Docker alternative for the local developer experience on my projects as I don't need all the functionalities provided by Docker Desktop. In my approach, I use the Docker engine, Docker CLI, and Docker Compose to spin up various technologies on the developers' machines for development purposes. 
Thanks to Rancher Desktop this approach is still possible as it offers out-of-the-box Docker open-source functionalities.
Using Rancher Desktop for my workflow comes at a cost as I have few issues with it.

1. Privileged mode - Rancher Desktop needs to be run as an Administrator on Windows to fully support containers' networking. As a consequence, every client of the Docker Engine installed with Rancher Desktop needs to run with the highest privileges as well. So IDEs or Terminals need to run as Administrator to interact with the Docker engine.
2. Starting Rancher Desktop at Windows startup as Administrator seems not to be possible. I tried on two different machines with Windows 10 by using different approaches (Task Manager, Windows Registry, Windows Startup folder) and it doesn't work.

Do you have any recommendations or advice for me?

**Comment by mook-as** [comment]
Hmm.
1. Privileged mode — as far as I recall, while we elevate to install Rancher Desktop (the _Rancher Desktop Privileged Service_), you should still be running the application as an unprivileged user.  It will communicate with the privileged service (because it's been installed to [allow interactive users to communicate with it](https://github.com/rancher-sandbox/rancher-desktop/blob/v1.9.1/scripts/lib/installer-win32-gen.tsx#L251)).
2. Because of that, starting Rancher Desktop at Window startup should be done as a non-administrator.

Could you expand on what is not working when you install Rancher Desktop privileged, and then run it as a normal (interactive) user? That may be a bug we need to fix. Thanks!

**Comment by enrico23** [comment]
Thanks for your suggestion.

This is my situation:

I installed Rancher Desktop "for everyone" on both Windows PCs that I mentioned in my previous post and by following the instructions on https://docs.rancherdesktop.io/faq, I added Rancher Desktop icon to the Windows startup.  I restarted the machine and Rancher ran at Windows Startup as expected.

Rancher Desktop is set to run as a "background service". In the preferences dialog box, under WSL - Integrations I selected the option "Expose Rancher Desktop's Kubernetes configuration and Docker socket to Windows Subsystem for Linux (WSL) distros"  and selected "Ubuntu" which is one of the distros at my disposal and is the default.
If I perform the command "wsl --list" on the machine I can see the following situation:

```
Windows Subsystem for Linux Distributions:
Ubuntu (Default)
rancher-desktop-data
rancher-desktop
```

If I open Ubuntu on WSL2 and type "docker info" everything seems to be working as expected apart from the "docker-credential-secretservice".  Docker-compose generates errors in getting credentials. This can be fixed by installing some packages.

If I use Git Bash which is based on a small distro of Linux or Powershell, and I type "docker info", I get the following error:
"Get "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/containers/json?all=1&filters=%7B%22label%22%3A%7B%22com.docker.compose.config-hash%22%3Atrue%2C%22com.docker.compose.project%3Dbuild%22%3Atrue%7D%7D": open //./pipe/docker_engine: Access is denied."

If I run Git Bash or Powershell as an Administrator everything works fine.


Just for comparison, when I was using Docker Desktop all the clients were working fine with the unprivileged user.

Thanks for your help.

**Comment by enrico23** [comment]
I overcame some of the limitations that encountered by modifying the configuration of the Ubuntu distro that I installed on WSL2.

My setup:
* Windows 10 
* WSL2
* Ubuntu 22.04.2 LTS

I just wanted to share my experience with other users and verify my assumptions:

1. Fix for DHCP Issue (resolv.config)
Most of the the times connection to the Internet is not immediately possible on Ubuntu. 
To fix this issue, open your Ubuntu on your favorite terminal and type the following commands:

```
sudo rm /etc/resolv.conf
sudo bash -c 'echo "nameserver 8.8.8.8" > /etc/resolv.conf'
sudo bash -c 'echo "[network]" > /etc/wsl.conf'
sudo bash -c 'echo "generateResolvConf = false" >> /etc/wsl.conf'
sudo chattr +i /etc/resolv.conf
```

2. Fix for denied access to the Docker daemon
The Docker engine is installed as part of Rancher Desktop. After the installation, access to the Docker deamon is normally denied with the following error message: `... open //./pipe/docker_engine: Access is denied...`
To fix this issue, open your Ubuntu on your favorite terminal and type the following commands:

```
### create docker group ###
sudo groupadd docker

### add user to the docker group ###
sudo usermod -aG docker $USER

### verify group assignment ###
groups user

### Adjust Docker Socket Permissions ###
sudo chmod 666 /var/run/docker.sock
```

**Comment by mook-as** [comment]
> Rancher Desktop is set to run as a "background service".

I assume this means that
- Rancher Desktop's shortcut is added to the Startup Start Menu folder (as documented [here](https://docs.rancherdesktop.io/faq/#q-how-can-i-add-rancher-desktop-to-the-startup-programs-list-on-windows))
- In Preferences, Rancher Desktop is set to start in the background, as in:
![image](https://github.com/rancher-sandbox/rancher-desktop/assets/3977982/f7e0aadb-ba62-4edd-bfca-c7340cc48c9b)

That should still make `docker.exe info` work… (Git Bash would be using msys; as long as you don't have a msys-native `git` executable, it should end up using the Win32 one and work as a normal user.  But some file path things would have issues.)

As for Ubuntu (in WSL2), that shouldn't be using `//./pipe/…` at all; that should be using `docker` that we ship, and contact `/var/run/docker.sock` (as long as your user is either `root` or a member of the `docker` group, which you did in your follow-up).  Restarting Rancher Desktop would reassign the group correctly.

That is, as far as I can tell, the original issue described (from _Windows_, `docker.exe info` isn't working because `\\.\pipe\docker_engine` is not accessible unless it's run in an elevated command prompt) isn't being resolved? I'd like to figure out the details on that so we can fix it…

**Reply by enrico23** [reply]
Today my plan is to reset Rancher Desktop to defaults, uninstall it, and do the same for Ubuntu 22.04. 
I will restart from scratch following your suggestion.

**Comment by enrico23** [comment]
Thanks for your suggestions. 

I have solved my issues and found a way to use the unprivileged user for my local DX workflows.

1) I have installed Rancher Desktop only for the current user with the "Install just for you" option. With this option, Rancher Desktop is installed only on the specified user account, but you will not need local Administrator privileges. This will also disable support for listening on non-loopback ports and you will not be able to expose applications/services on any interface except 127.0.0.1.

2) I set the following preferences:

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/16436149/990c79dd-6a1f-4454-9559-53af7866b501)

After reboot, I could see the Rancher Desktop icon in the system tray:

![image](https://github.com/rancher-sandbox/rancher-desktop/assets/16436149/581a3398-0962-4d72-9d08-0d4b3c063fe3)


With this approach, the only change that I had to make was on the connection strings used for the various integrations running on containers. I couldn't use any more 'localhost' in them. I had to use 127.0.01 instead.

Thanks again and keep up the great work!
